files/migrations.py

```diff
@@ -0,0 +1,188 @@
+#!/usr/bin/python3
+"""
+Runs migration scripts for this project
+
+Intended to be triggered by the %post rpm scriptlet rather than manually.
+"""
+
+import logging as log
+import shutil
+import sys
+import tempfile
+from importlib.machinery import SourceFileLoader
+from importlib.util import module_from_spec, spec_from_loader
+from pathlib import Path
+
+
+class Version:
+    """
+    A sortable software version
+
+    Reads version strings, and sorts by major.minor.patch[.etc[.etc …]]
+    """
+
+    def __init__(self, version):
+        """
+        Read version numbers into lists of ints, discard anything that comes
+        after the first character that is neither a period or a digit
+        """
+        chunks = []
+        # Opting for minimal code over execution speed
+        for part in version.split("."):
+            digits = ""
+            for char in part:
+                if char.isdigit():
+                    digits += char
+                else:
+                    # Treat RCs, betas etc. as if it were the full release
+                    break
+            chunks.append(int(digits))
+        self.version = tuple(chunks)
+
+    def __str__(self):
+        return ".".join(str(v) for v in self.version)
+
+
+class Migration(Version):
+    """
+    A sortable runnable migration
+
+    Subclass of Version so that we can sort migrations just as easily as version strings
+    """
+
+    def __init__(self, path):
+        super().__init__(path.name.rsplit(".", 1)[0])
+        self.path = path
+
+    def run_and_update_version_file(self, version_file):
+        """
+        Run this migration
+        """
+
+        try:
+            loader = SourceFileLoader("migration", str(self.path))
+            spec = spec_from_loader("migration", loader)
+            migration = module_from_spec(spec)
+            loader.exec_module(migration)
+            log.info(f"Running migration for {self}")
+            migrate(migration.steps)
+        except Exception as error:
+            log.error(f"{error}")
+            log.error(f"Failed to load migration: {self.path}")
+            exit(2)
+
+        # If we successfully ran the migration, we are now in a new state that we want to see
+        # reflected in case a subsequent migration fails
+        update_version(self, version_file)
```

Comment:
Re: subclassing, I think seeing `update_version(self.version, version_file)` would actually be clearer than passing `self` and it implicitly calling `self.__str__()` via the f-string interpolation.